### PR TITLE
[complex-half] support casting (by updating copy_)

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference https://github.com/pytorch/pytorch/issues/71680

### CODE REVIEW

**Comment by anjali411** [comment]
looks good to me but needs testing. we can likely just extend the opinfo tests for complex half?

more generally, I think it might be nice to create a tracker and a bunch of tasks/functions that we are looking to add complex half support for. It would be really nice for bookkeeping and to keep track of things.

**Comment by kshitij12345** [comment]
> I think it might be nice to create a tracker and a bunch of tasks/functions that we are looking to add complex half support for. It would be really nice for bookkeeping and to keep track of things.

Agreed. Will update the [issue](https://github.com/pytorch/pytorch/issues/71680) soon (in a day or two) with the list of ops.

File: c10/util/TypeCast.h [hunk 1]
```diff
@@ -68,6 +69,36 @@ struct static_cast_with_inter_type<uint8_t, src_t> {
   }
 };
 
+template <>
+struct static_cast_with_inter_type<c10::complex<c10::Half>, c10::BFloat16> {
+  C10_HOST_DEVICE __ubsan_ignore_undefined__ static inline c10::complex<
+      c10::Half>
+  apply(c10::BFloat16 src) {
+    return static_cast<c10::complex<c10::Half>>(c10::complex<float>{src});
```
Comment (anjali411) [inline c10/util/TypeCast.h:77]:
```suggestion
    return static_cast<c10::complex<c10:: BFloat16 >>(c10::complex<float>{src});
```

File: c10/util/TypeCast.h [hunk 2]
```diff
@@ -68,6 +69,36 @@ struct static_cast_with_inter_type<uint8_t, src_t> {
...
+};
+
+template <>
+struct static_cast_with_inter_type<
+    c10::complex<c10::Half>,
+    c10::complex<double>> {
+  C10_HOST_DEVICE __ubsan_ignore_undefined__ static inline c10::complex<
+      c10::Half>
+  apply(c10::complex<double> src) {
+    return static_cast<c10::complex<c10::Half>>(
```
Comment (anjali411) [inline c10/util/TypeCast.h:97]: why can't we directly cast complex half to complex double?

File: test/test_torch.py [hunk 3]
```diff
@@ -5299,6 +5300,43 @@ def test_assertRaisesRegex_ignore_msg_non_native_device(self, device):
...
+            src = make_tensor_wrapper((50,), dtype=src_dtype)
+            t.copy_(src)
+            dst = make_tensor_wrapper((50, ), dtype=src_dtype)
+            if can_cast(src_dtype, dtype):
+                rtol = None
+                atol = None
+                if dtype in (torch.half, torch.complex32):
+                    rtol = 1e-3
+                    atol = 1e-3
+                if dtype in (torch.bfloat16,):
```
Comment (anjali411) [inline test/test_torch.py:5335]: we should be able to update this after the fix above?

**Comment by anjali411** [comment]
we should probably add a complex32 label. also I'd recommend adding a follow-up to add copy_ OpInfo after we support the minimal set of operators to be able to do that

**Comment by anjali411** [comment]
LGTM once the CI is green

File: test/test_torch.py [hunk 4]
```diff
@@ -5299,6 +5300,43 @@ def test_assertRaisesRegex_ignore_msg_non_native_device(self, device):
         with self.assertRaisesRegex(RuntimeError, msg):
             torch.nn.functional.nll_loss(x, t, weight=invalid_weight)
 
+    @dtypes(*all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16, torch.complex32))
+    def test_copy_(self, device, dtype):
```
Comment (anjali411) [inline test/test_torch.py:5304]: there are a bunch of tests for copy in `test_torch.py` for e.g., `test_copy_all_dtypes_and_devices` we should run the existing tests for complex32 wherever possible

File: c10/util/TypeCast.h (hunk 1 above)
Comment (kshitij12345) [reply c10/util/TypeCast.h:77]: It should be `c10::complex<c10::Half>`, right? Our complex32 will correspond to only that. So this will allow casting from `BFloat16` to `complex<Half>`.

File: c10/util/TypeCast.h (hunk 2 above)
Comment (kshitij12345) [reply c10/util/TypeCast.h:97]: Currently, we don't have a constructor for `complex<Half>` which takes `complex<double>`. Having that constructor leads to ambiguos resolution while converting from `double` to `complex<Half>` reason being both constructors which take `complex<float>` and `complex<double>` will be equally valid (as both `complex<float>` and `complex<double>` can be constructed from a double).

```
../c10/util/TypeCast.h: In instantiation of 'static dest_t c10::static_cast_with_inter_type<dest_t, src_t>::apply(src_t) [with dest_t = c10::complex<c10::Half>; src_t = double]':
../c10/util/TypeCast.h:212:54:   required from 'To c10::convert(From) [with To = c10::complex<c10::Half>; From = double]'
../c10/util/TypeCast.h:224:27:   required from 'To c10::checked_convert(From, const char*) [with To = c10::complex<c10::Half>; From = double]'
../c10/core/Scalar.h:70:3:   required from here
../c10/util/TypeCast.h:49:12: error: call of overloaded 'complex(double&)' is ambiguous
   49 |     return static_cast<dest_t>(r);
      |            ^~~~~~~~~~~~~~~~~~~~~~
In file included from ../c10/core/ScalarType.h:5,
                 from ../c10/core/Scalar.h:11,
                 from aten/src/ATen/core/TensorBody.h:16,
                 from ../aten/src/ATen/core/jit_type.h:5,
                 from ../aten/src/ATen/core/function_schema.h:6,
                 from ../aten/src/ATen/core/op_registration/infer_schema.h:8,
                 from ../torch/library.h:61,
                 from ../aten/src/ATen/BatchingRegistrations.cpp:1:
../c10/util/Half.h:409:35: note: candidate: 'c10::complex<c10::Half>::complex(const c10::complex<double>&)'
  409 |   C10_HOST_DEVICE explicit inline complex(const c10::complex<double>& value)
      |                                   ^~~~~~~
../c10/util/Half.h:407:35: note: candidate: 'c10::complex<c10::Half>::complex(const c10::complex<float>&)'
  407 |   C10_HOST_DEVICE explicit inline complex(const c10::complex<float>& value)
      |                                   ^~~~~~~
```

I am also planning to look at all the constructors which `complex<Half>` should have. Currently it also lacks a constructor from `Half` to `complex<Half>` i.e. from real to complex.

File: test/test_torch.py (hunk 3 above)
Comment (kshitij12345) [reply test/test_torch.py:5335]: I don't think so. The test casts from `src_t` -> `dst_t` -> `src_t` and compares the original `src_t` to the reconstructed `src_t`.
So if `src_t` is float and `dst_t` is BFloat16 or Half, then there would be precision loss from the original data and the reconstructed `src_t` will also have the same. So we need to relax the tolerance.

Also note that `bfloat16` case will be hit only when `src_t` is float kind not complex (as we have this check only when `src_t` and `dst_t` are of same kind)

Let me know if that made sense.

Thanks!

File: test/test_torch.py (hunk 4 above)
Comment (kshitij12345) [reply test/test_torch.py:5176]: I couldn't update a bunch of test since they use unsupported ops on `complex32` (I updated the few I could).

Will track the ones which weren't updated in tracker issue

Updated Tests:
* test_clone_all_dtypes_and_devices
* test_copy_dtypes
* test_copy_transpose

Not updated Tests:
* test_copy_all_dtypes_and_devices -> Fails due to copy(x) (unsupported storage type )
* test_copy_math_view -> Uses make_tensor
* test_copy_transpose_math_view -> Uses make_tensor
* test_copy_noncontig -> Tests only 1 dtype
* test_copy_broadcast -> Tests only 1 dtype

File: c10/util/TypeCast.h (hunk 2 above)
Comment (anjali411) [reply c10/util/TypeCast.h:97]: oh cool thanks for clarifying!

> I am also planning to look at all the constructors which complex<Half> should have. Currently it also lacks a constructor from Half to complex<Half> i.e. from real to complex.

good idea! a good thing to add in the tracker as well

File: c10/util/TypeCast.h (hunk 1 above)
Comment (anjali411) [reply c10/util/TypeCast.h:77]: you are totally right I got confused

File: test/test_torch.py (hunk 3 above)
Comment (anjali411) [reply test/test_torch.py:5207]: > Also note that bfloat16 case will be hit only when src_t is float kind not complex (as we have this check only when src_t and dst_t are of same kind)

I don't follow this but everything else makes sense

**Review by anjali411 — APPROVED** [review]

File: test/test_torch.py (hunk 4 above)
Comment (anjali411) [reply test/test_torch.py:5176]: sounds good!
